core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -290,8 +292,10 @@ private int readVersionHint() {
         return Integer.parseInt(in.readLine().replace("\n", ""));
       }
 
-    } catch (IOException e) {
-      throw new RuntimeIOException(e, "Failed to get file system for path: %s", versionHintFile);
+    } catch (Exception e) {
+      LOG.warn("Error reading version hint file {}", versionHintFile, e);
+      // We just assume corrupted metadata and start to read from the first version file
+      return 1;
```

Comment:
Should this return `0` instead? Similar to when the version hint file does not exist?

Reply:
When we are returning 0 the caller will end up throwing the following exception:
```
Table does not exist: tbl
org.apache.iceberg.exceptions.NoSuchTableException: Table does not exist: tbl
	at org.apache.iceberg.BaseMetastoreCatalog.loadTable(BaseMetastoreCatalog.java:108)
	at org.apache.iceberg.hadoop.TestHadoopCatalog.testVersionHintFile(TestHadoopCatalog.java:508)
```

The goal of the PR is to recover from the corrupted version-hint.txt file, so I have returned 1 instead.
But this question made me wonder if returning 1 is a good idea or not, so I will check with Ryan

Reply:
@rdblue: I have checked @shardulm94's comment and started to think about the possible problems/solutions.
The main question is:
- Are the metadataFiles kept forever? Or are they cleaned-up with compaction, or whatever other process?

If they are kept forever then it could be ok to start from 1. If they could be removed somehow then we might end up failing to read the removed version file and failing anyway. Could this be a problem? Could it be a good solution to list the metadata files for this case and trying to recover using that listing?

Thanks,
Peter

Reply:
0 is currently used to signal that the version hint doesn't exist. The `refresh`  method checks whether the version hint was missing so that it can throw an exception when the metadata is missing: if there is no metadata file and the version hint was 0, then the table doesn't exist yet. If the version hint is non-zero, then the problem is that the metadata file is missing.

I like the solution here that does some recovery to find the right version. We should probably rename the method since it is doing recovery, but it's fine for now.

The recovery that is done in `readVersionHint` is something we can improve on. We could use file listing if the v1 metadata file doesn't exist. That would be a fairly safe way of recovering the version hint. I think that means there should be 2 follow-ups:

1. Improve how we write the hint to avoid corruption in HDFS by writing a file to a unique name, then renaming it to `version-hint.txt`. That requires overwriting with the rename, so I'm not sure if we can actually do this.
2. Improve version recovery when metadata files have been cleaned up, probably using list operations.

Reply:
#1465 is created for the 2nd point